File: utils/data_processing.py
import pandas as pd
import numpy as np
from datetime import datetime

def validate_csv(df):
    """
    Validate that CSV has required columns and proper data types.

    Args:
        df: pandas DataFrame

    Returns:
        tuple: (is_valid, error_message)
    """
    required_columns = [
        'meetingDate', 'Darkstore', 'traderName', 'traderId',
        'buyerName', 'buyerid', 'buyerPhone',
        'currentLatitude', 'currentLongitude',
        'selfie', 'verificationDoc'
    ]

    missing_columns = [col for col in required_columns if col not in df.columns]

    if missing_columns:
        return False, f"Missing required columns: {', '.join(missing_columns)}"

    return True, None


def clean_and_process_data(df):
    """
    Clean and process the uploaded CSV data.

    Args:
        df: pandas DataFrame

    Returns:
        pandas DataFrame: Cleaned data
    """
    df = df.copy()

    # Convert coordinates to numeric, handle errors
    df['currentLatitude'] = pd.to_numeric(df['currentLatitude'], errors='coerce')
    df['currentLongitude'] = pd.to_numeric(df['currentLongitude'], errors='coerce')

    # Remove rows with invalid GPS coordinates
    initial_count = len(df)
    df = df.dropna(subset=['currentLatitude', 'currentLongitude'])
    removed_count = initial_count - len(df)

    if removed_count > 0:
        print(f"Removed {removed_count} rows with invalid GPS coordinates")

    # Validate GPS coordinate ranges
    df = df[
        (df['currentLatitude'].between(-90, 90)) &
        (df['currentLongitude'].between(-180, 180))
    ]

    # Parse meeting date
    df['meetingDate'] = pd.to_datetime(df['meetingDate'], errors='coerce')

    # Clean phone numbers (remove non-digits)
    df['buyerPhone'] = df['buyerPhone'].astype(str).str.replace(r'\D', '', regex=True)

    # Handle NULL/empty selfie and verification doc URLs
    df['selfie'] = df['selfie'].fillna('').astype(str)
    df['verificationDoc'] = df['verificationDoc'].fillna('').astype(str)

    # Remove brackets and quotes from selfie column (handle "[]" cases)
    df['selfie'] = df['selfie'].str.replace(r'[\[\]"\']', '', regex=True)
    df['verificationDoc'] = df['verificationDoc'].str.replace(r'[\[\]"\']', '', regex=True)

    # Clean buyer and trader names
    df['buyerName'] = df['buyerName'].astype(str).str.strip()
    df['traderName'] = df['traderName'].astype(str).str.strip()
    df['Darkstore'] = df['Darkstore'].astype(str).str.strip()

    # Add processed flag
    df['has_valid_gps'] = True
    df['has_selfie'] = df['selfie'].str.len() > 0
    df['has_verification'] = df['verificationDoc'].str.len() > 0

    return df


def get_filter_options(df):
    """
    Extract unique values for filters.

    Args:
        df: pandas DataFrame

    Returns:
        dict: Dictionary with filter options
    """
    return {
        'darkstores': sorted(df['Darkstore'].unique().tolist()),
        'traders': sorted(df['traderName'].unique().tolist()),
        'date_range': (df['meetingDate'].min(), df['meetingDate'].max())
    }


def apply_filters(df, darkstores=None, traders=None, date_range=None):
    """
    Apply filters to the dataframe.

    Args:
        df: pandas DataFrame
        darkstores: list of darkstore names to include
        traders: list of trader names to include
        date_range: tuple of (start_date, end_date)

    Returns:
        pandas DataFrame: Filtered data
    """
    filtered_df = df.copy()

    if darkstores:
        filtered_df = filtered_df[filtered_df['Darkstore'].isin(darkstores)]

    if traders:
        filtered_df = filtered_df[filtered_df['traderName'].isin(traders)]

    if date_range and len(date_range) == 2:
        start_date, end_date = date_range
        filtered_df = filtered_df[
            (filtered_df['meetingDate'] >= pd.Timestamp(start_date)) &
            (filtered_df['meetingDate'] <= pd.Timestamp(end_date))
        ]

    return filtered_df

File: lib/dataProcessing.ts
import type { Row, Filters, FilterOptions } from './types'

const REQUIRED_COLUMNS = [
  'meetingDate', 'Darkstore', 'traderName', 'traderId',
  'buyerName', 'buyerid', 'buyerPhone',
  'currentLatitude', 'currentLongitude',
  'selfie', 'verificationDoc',
]

export function validateCSV(rows: Record<string, string>[]): { valid: boolean; error?: string } {
  if (rows.length === 0) return { valid: false, error: 'CSV file is empty' }
  const cols = Object.keys(rows[0])
  const missing = REQUIRED_COLUMNS.filter(c => !cols.includes(c))
  if (missing.length > 0) return { valid: false, error: `Missing required columns: ${missing.join(', ')}` }
  return { valid: true }
}

export function cleanAndProcessData(raw: Record<string, string>[]): Row[] {
  const rows: Row[] = []

  for (const r of raw) {
    const lat = parseFloat(r.currentLatitude)
    const lon = parseFloat(r.currentLongitude)

    // Skip invalid coordinates
    if (isNaN(lat) || isNaN(lon)) continue
    if (lat < -90 || lat > 90 || lon < -180 || lon > 180) continue

    const selfie = (r.selfie ?? '').replace(/[\[\]'"]/g, '').trim()
    const verificationDoc = (r.verificationDoc ?? '').replace(/[\[\]'"]/g, '').trim()
    const phone = (r.buyerPhone ?? '').replace(/\D/g, '')

    rows.push({
      meetingDate: r.meetingDate ?? '',
      meetingTime: r.meetingTime,
      Darkstore: (r.Darkstore ?? '').trim(),
      traderName: (r.traderName ?? '').trim(),
      traderId: r.traderId ?? '',
      buyerName: (r.buyerName ?? '').trim(),
      buyerid: r.buyerid ?? '',
      buyerPhone: phone,
      buyerRegistrationDate: r.buyerRegistrationDate,
      currentLatitude: lat,
      currentLongitude: lon,
      selfie,
      verificationDoc,
      remarks: r.remarks,
      distancemtr: r.distancemtr,
      todayOrders: r.todayOrders,
      has_valid_gps: true,
      has_selfie: selfie.length > 0,
      has_verification: verificationDoc.length > 0,
    })
  }

  return rows
}

export function getFilterOptions(rows: Row[]): FilterOptions {
  const darkstores = [...new Set(rows.map(r => r.Darkstore))].filter(Boolean).sort()
  const traders = [...new Set(rows.map(r => r.traderName))].filter(Boolean).sort()
  const dates = rows.map(r => r.meetingDate).filter(Boolean).sort()
  return {
    darkstores,
    traders,
    dateRange: [dates[0] ?? '', dates[dates.length - 1] ?? ''],
  }
}

export function applyFilters(rows: Row[], filters: Filters): Row[] {
  let filtered = rows

  if (filters.darkstores.length > 0) {
    filtered = filtered.filter(r => filters.darkstores.includes(r.Darkstore))
  }

  if (filters.traders.length > 0) {
    filtered = filtered.filter(r => filters.traders.includes(r.traderName))
  }

  if (filters.dateRange) {
    const [start, end] = filters.dateRange
    filtered = filtered.filter(r => r.meetingDate >= start && r.meetingDate <= end)
  }

  if (filters.lastMeetingAfter) {
    // Find last meeting date per buyerid+traderId pair
    const lastMeeting: Record<string, string> = {}
    for (const r of filtered) {
      const key = `${r.buyerid}::${r.traderId}`
      if (!lastMeeting[key] || r.meetingDate > lastMeeting[key]) {
        lastMeeting[key] = r.meetingDate
      }
    }
    const pairsToKeep = new Set(
      Object.entries(lastMeeting)
        .filter(([, date]) => date >= filters.lastMeetingAfter!)
        .map(([key]) => key)
    )
    filtered = filtered.filter(r => pairsToKeep.has(`${r.buyerid}::${r.traderId}`))
  }

  return filtered
}
